modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/GridCacheDatabaseSharedManager.java

```diff
@@ -3276,31 +3296,28 @@ private void finalizeCheckpointOnRecovery(
 
                 writePageBuf.order(ByteOrder.nativeOrder());
 
-                Collection<FullPageId> pages0 = pages.innerCollection(innerIdx);
-
-                FullPageId fullPageId = null;
+                T2<PageMemoryEx, FullPageId> curPageId = pages.poll();
 
                 try {
-                    for (FullPageId fullId : pages0) {
+                    while (curPageId != null) {
                         // Fail-fast break if some exception occurred.
                         if (writePagesError.get() != null)
                             break;
 
-                        // Save pageId to local variable for future using if exception occurred.
-                        fullPageId = fullId;
-
-                        PageMemoryEx pageMem = getPageMemoryForCacheGroup(fullId.groupId());
+                        PageMemoryEx pageMem = curPageId.get1();
 
                         // Write page content to page store via pageStoreWriter.
                         // Tracker is null, because no need to track checkpoint metrics on recovery.
-                        pageMem.checkpointWritePage(fullId, writePageBuf, pageStoreWriter, null);
-                    }
+                        pageMem.checkpointWritePage(curPageId.get2(), writePageBuf, pageStoreWriter, null);
 
-                    // Add number of handled pages.
-                    cpPagesCnt.addAndGet(pages0.size());
+                        curPageId = pages.poll();
+
+                        // Add number of handled pages.
+                        cpPagesCnt.incrementAndGet();
```

Comment:
Maybe it's better to increment global cpPagesCnt only one time at the end of the thread execution but inside of the thread using the local counter?